viogpu/viogpu3d/viogpu_device.cpp

```diff
@@ -0,0 +1,337 @@
+        UINT rectCnt = min(pPresent->SubRectCnt, (pPresent->DmaSize - 0x100) / sizeOfOneRect);
+        
+        VIOGPU_COMMAND_HDR* cmd_hdr = (VIOGPU_COMMAND_HDR*)dmaBuf;
+        cmd_hdr->type = VIOGPU_CMD_SUBMIT;
+        cmd_hdr->size = rectCnt * sizeOfOneRect;
+        dmaBuf += sizeof(VIOGPU_COMMAND_HDR);
+
+        for (UINT i = 0; i < rectCnt; i++) {
+            UINT* cmdBody = (UINT*)dmaBuf;
+            dmaBuf += sizeOfOneRect;
+
+            RECT rect = pPresent->pDstSubRects[i];
+
+            cmdBody[0] = VIRGL_CMD0(VIRGL_CCMD_RESOURCE_COPY_REGION, 0, VIRGL_CMD_RESOURCE_COPY_REGION_SIZE);
+            cmdBody[1] = dst->GetId();
+            cmdBody[2] = 0;
+            cmdBody[3] = rect.left;
+            cmdBody[4] = rect.top;
+            cmdBody[5] = 0;
+
+            cmdBody[6] = src->GetId();
+            cmdBody[7] = 0;
+            cmdBody[8] = rect.left + dx;
+            cmdBody[9] = rect.top + dy;
+            cmdBody[10] = 0;
+            cmdBody[11] = rect.right - rect.left;
+            cmdBody[12] = rect.bottom - rect.top;
+            cmdBody[13] = 1;
+        }
+    }
+
+    if(dst->IsCoherent()) {
+        VIOGPU_COMMAND_HDR* cmd_hdr = (VIOGPU_COMMAND_HDR*)dmaBuf;
+        cmd_hdr->type = VIOGPU_CMD_TRANSFER_FROM_HOST;
+        cmd_hdr->size = sizeof(VIOGPU_TRANSFER_CMD);
+        dmaBuf += sizeof(VIOGPU_COMMAND_HDR);
+
+        VIOGPU_TRANSFER_CMD* cmdBody = (VIOGPU_TRANSFER_CMD*)dmaBuf;
+        dmaBuf += sizeof(VIOGPU_TRANSFER_CMD);
+
+        cmdBody->res_id = dst->GetId();
+
+        cmdBody->box.x = coverRect.left;
+        cmdBody->box.y = coverRect.top;
+        cmdBody->box.z = 0;
+        cmdBody->box.width = coverRect.right - coverRect.left;
+        cmdBody->box.height = coverRect.bottom - coverRect.top;
+        cmdBody->box.depth = 1;
+
+        cmdBody->layer_stride = 0;
+        cmdBody->stride = 0;
+        cmdBody->level = 0;
+        cmdBody->offset = 0;
+    }
+
+    pPresent->pDmaBuffer = dmaBuf;
+
+    return STATUS_SUCCESS;
+}
+
+NTSTATUS VioGpuDevice::Present(_Inout_ DXGKARG_PRESENT* pPresent)
+{
+    PAGED_CODE();
+
+    if (pPresent->Flags.Flip) return STATUS_SUCCESS;
+
+    DbgPrint(
+        TRACE_LEVEL_VERBOSE,
+        ("<---> %s Flags=%s %s %s %s %s %s %s %s)\n",
+            __FUNCTION__,
+            pPresent->Flags.Blt ? "Blt" : "",
+            pPresent->Flags.ColorFill ? "ColorFill" : "",
+            pPresent->Flags.Flip ? "Flip" : "",
+            pPresent->Flags.FlipWithNoWait ? "FlipWithNoWait" : "",
+            pPresent->Flags.SrcColorKey ? "SrcColorKey" : "",
+            pPresent->Flags.DstColorKey ? "DstColorKey" : "",
+            pPresent->Flags.LinearToSrgb ? "LinearToSrgb" : "",
+            pPresent->Flags.Rotate ? "Rotate" : ""));
+
+    VioGpuCommand* cmd = new(NonPagedPoolNx) VioGpuCommand(m_pAdapter);
+    if (pPresent->pDmaBuffer) {
+        VioGpuCommand** privateData = (VioGpuCommand**)pPresent->pDmaBufferPrivateData;
+        *privateData = cmd;
+    }
```

Comment:
Is this leaked otherwise?

Reply:
Not 100% sure, but `pPresent->pDmaBuffer` is set there (so the condition is always true). `pPresent->pDmaBuffer` afair is not set only when `FlipOnVSyncMmIo` is set and Present is called with `pPresent->Flags.Flip`, but we check for that earlier so `pPresent->pDmaBuffer` should be valid there. I will check a bit more and if it is correct the condition will be removed.